storage: Unflake TestRaftHeartbeats

rjnn opened the pull request:
The previous structure of TestRaftHeartbeats would occasionally fail
if the initial leader was a replica not on node 1. Fix this by
recording the initial leader (wherever it may be) and just looking for
that leader to stay constant after a raft of heartbeats.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10277)

<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 1439 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10277#-KV7gBu58T3ZeCRsl0qz:-KV7gBu58T3ZeCRsl0r-:b-f2jmzi) ([raw file](https://github.com/cockroachdb/cockroach/blob/f072f4cc2f6a85a3eaf2872feeb4134645f848bb/pkg/storage/client_raft_test.go#L1439)):*
> ```Go
> 		if mtc.stores[i].RaftStatus(1).SoftState.RaftState == raft.StateLeader {
> 			leaderID = i
> 			initialTerm = mtc.stores[i].RaftStatus(1).Term
> ```

You could initialize this outside of the loop:

```
  initialTerm := mtc.stores[leaderID].RaftStatus(1).Term
```

Also: `s/leaderID/leaderIdx/g` as `i` is an index into `mtc.stores`, not the ID of the store.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10277#-:-KV7gUigi1nNYC3uAjZh:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 1436 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10277#-KVAkirg6BLJRbnTRdTl:-KVAkirg6BLJRbnTRdTm:b-fd3k55) ([raw file](https://github.com/cockroachdb/cockroach/blob/f072f4cc2f6a85a3eaf2872feeb4134645f848bb/pkg/storage/client_raft_test.go#L1436)):*
> ```Go
> 	leaderID := -1
> 	var initialTerm uint64
> 	for i := 0; i < 3; i++ {
> ```

`for i := range mtc.stores`? I would really love it if we stopped instinctively typing `3` everywhere as if it was a constant enshrined by the cockroach gods.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10277)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 1436 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10277#-KVAkirg6BLJRbnTRdTl:-KVDAqMIgcx0H3VfOZ-3:b-qp7il2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f072f4cc2f6a85a3eaf2872feeb4134645f848bb/pkg/storage/client_raft_test.go#L1436)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `for i := range mtc.stores`? I would really love it if we stopped instinctively typing `3` everywhere as if it was a constant enshrined by the cockroach gods.
</details>
Sorry, was following scripture. https://youtu.be/xOrgLj9lOwk?t=89

Fixed, and noted for the future. Thank you.

---

*[pkg/storage/client_raft_test.go, line 1439 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10277#-KV7gBu58T3ZeCRsl0qz:-KVDBphAZmARpyiXszSc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f072f4cc2f6a85a3eaf2872feeb4134645f848bb/pkg/storage/client_raft_test.go#L1439)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> You could initialize this outside of the loop:
> 
> ```
>   initialTerm := mtc.stores[leaderID].RaftStatus(1).Term
> ```
> 
> Also: `s/leaderID/leaderIdx/g` as `i` is an index into `mtc.stores`, not the ID of the store.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10277)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10277)*
<!-- Sent from Reviewable.io -->

Commit:
storage: Unflake TestRaftHeartbeats

The previous structure of TestRaftHeartbeats would occasionally fail
if the initial leader was a replica not on node 1. Fix this by
recording the initial leader (wherever it may be) and just looking for
that leader to stay constant after a raft of heartbeats. Also change
the wait to a function of the tick interval by measuring the actual
ticks that have advanced, rather than a fixed duration sleep.

arjunravinarayan commented:
I adapted the tick wait code from #10214 here as well

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10277#-:-KVL3EDyGik-Pvcdp4oE:b-hsgenw)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10277#-:-KVLd-Ea0S3qxLjqnYOa:bnfp4nl)*
<!-- Sent from Reviewable.io -->
